Towards #2709 - Allow port instead of portIndex for health checks

kolloch opened the pull request:
* Improved health check documentation
  * Replaced JInt with scala Int in HealthCheck case class
  * Automatically add `"portIndex": 0` when necessary to preserve backwards compatibility

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala:
```diff
@@ -446,6 +447,26 @@ trait V2Formats {
+    val portIndexesMakeSense = v2App.ports.nonEmpty || hasPortMappings
+    if (portIndexesMakeSense) {
+      v2App.copy(healthChecks = v2App.healthChecks.map { healthCheck =>
+        if (healthCheck.port.isEmpty && healthCheck.portIndex.isEmpty && healthCheck.protocol != Protocol.COMMAND) {
```
Can you extract those checks as flag: `val needPortIndex = x&y&z ` and then have only one if: `if (needPortIndex && portIndexesMakeSense) app.copy(...)`

kolloch replied:
You mean like this?

```scala

  /**
    * Ensure backwards compatibility by adding portIndex to health checks when necessary.
    *
    * In the past, healthCheck.portIndex was required and had a default value 0. When we introduced healthCheck.port, we
    * made it optional (also with ip-per-container in mind) and we have to re-add it in cases where it makes sense.
    */
  private[this] def addHealthCheckPortIndexIfNecessary(v2App: V2AppDefinition): V2AppDefinition = {
    val hasPortMappings = v2App.container.exists(_.docker.exists(_.portMappings.exists(_.nonEmpty)))
    val portIndexesMakeSense = v2App.ports.nonEmpty || hasPortMappings
    v2App.copy(healthChecks = v2App.healthChecks.map { healthCheck =>
        def needsDefaultPortIndex = healthCheck.port.isEmpty && healthCheck.portIndex.isEmpty && healthCheck.protocol != Protocol.COMMAND
        if (portIndexesMakeSense && needsDefaultPortIndex) healthCheck.copy(portIndex = Some(0))
        else healthCheck
    })
  }
```

Slightly less efficient but more readable, I like it.

aquamatthias commented on src/main/scala/mesosphere/marathon/health/HealthCheck.scala:
```diff
@@ -54,7 +58,13 @@ case class HealthCheck(
+      portIndex =
+        if (proto.hasPortIndex)
+          Some(proto.getPortIndex)
+        else if (!proto.hasPort && proto.getProtocol != Protocol.COMMAND)
```
Can we run into this situation?

kolloch replied:
We are not sure. We might be for backwards compatibility. As written below, portIndex = 0 was the default int protobufs. We don't know if we can get into this situation by a combination of using an old Marathon version and migrations so we thought that we better cover it.

aquamatthias commented on src/test/scala/mesosphere/marathon/health/HealthCheckTest.scala:
```diff
@@ -31,12 +30,35 @@ class HealthCheckTest extends MarathonSpec {
+    val healthCheck = HealthCheck(
+      path = Some("/health"),
+      protocol = Protocol.HTTP,
+      portIndex = Some(0),
```
Isn't that something we forbid to create: portIndex and port?

kolloch replied:
Yes, we do that. This test only covers the serialization. Should we separate this to two different tests?

aquamatthias commented:
Only minor comments. Looks good! :+1: 
Please address and :ship:

kolloch commented:
@aquamatthias I hopefully addressed all your concerns in an additional commit. I can squash our two commits if you are happy (I would like to preserve Evans to give him credit where it's due).

aquamatthias commented:
Thanks! LGTM please squash and merge.

kolloch commented:
It is squashed! :)  We will not squash Evans patch.

aquamatthias commented:
Looks good. Thanks!
